pkg/agent/controller/networkpolicy/l7engine/reconciler.go

```diff
@@ -0,0 +1,398 @@
+		}
+	}
+
+	klog.Infof("Reconciling L7 rule %s of %s", ruleID, policyName)
+	// Write the Suricata rules to file.
+	rulesPath := generateRulesPath(vlanID)
+	rulesData := generateRulesConf(policyName, ipSrc, ipDst, tpDst, protoKeywords)
+	if err := writeConfigFile(rulesPath, rulesData); err != nil {
+		return fmt.Errorf("failed to write Suricata rules to file %s for L7 rule %s of %s", rulesPath, ruleID, policyName)
+	}
+
+	// Add a Suricata tenant.
+	if err := r.addSuricataTenant(vlanID, rulesPath); err != nil {
+		return fmt.Errorf("failed to add Suricata tenant for L7 rule %s of %s", ruleID, policyName)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) DeleteRule(ruleID string, vlanID uint32) error {
+	// Delete the Suricata tenant.
+	if err := r.deleteSuricataTenant(vlanID); err != nil {
+		return fmt.Errorf("failed to delete Suricata tenant %d for L7 rule %s: %v", vlanID, ruleID, err)
+	}
+
+	// Delete the Suricata rules file.
+	rulesPath := generateRulesPath(vlanID)
+	if err := defaultFS.Remove(rulesPath); err != nil {
+		return fmt.Errorf("failed to delete rules file %s for L7 rule %s: %v", rulesPath, ruleID, err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) addSuricataTenant(vlanID uint32, rulesPath string) error {
+	tenantConfigPath := generateTenantConfigPath(vlanID)
+	exists, err := afero.Exists(defaultFS, tenantConfigPath)
+	if err != nil {
+		return fmt.Errorf("failed to stat tenant config file %s", tenantConfigPath)
+	}
+
+	// If the tenant config file exists, it means that this tenant has been added, just reload the tenant to load updated
+	if exists {
+		scCmd := fmt.Sprintf("reload-tenant %d %s", vlanID, tenantConfigPath)
+		if output, err := r.suricataScFn(scCmd); err != nil {
+			return fmt.Errorf("failed to reload tenant %d with config file %s: %v", vlanID, tenantConfigPath, output)
+		}
+		return nil
+	}
+
+	// If the tenant config file doesn't exist, create a config file for the tenant.
+	tenantConfigData := bytes.NewBuffer([]byte(fmt.Sprintf(`%%YAML 1.1
+
+---
+default-rule-path: /etc/suricata/rules
+rule-files:
+  - %s
+`, rulesPath)))
+	if err = writeConfigFile(tenantConfigPath, tenantConfigData); err != nil {
+		return fmt.Errorf("failed to create config file %s for tenant %d: %v", tenantConfigPath, vlanID, err)
+	}
+
+	// Register the tenant with the config file. Note that, to be simple, use the VLAN id as the tenant ID.
+	scCmd := fmt.Sprintf("register-tenant %d %s", vlanID, tenantConfigPath)
+	if output, err := r.suricataScFn(scCmd); err != nil {
+		defer defaultFS.Remove(tenantConfigPath)
+		return fmt.Errorf("failed to register tenant %d with config file %s: %v", vlanID, tenantConfigPath, output)
+	}
+
+	// Register the tenant handler by mapping the tenant to the VLAN ID.
+	scCmd = fmt.Sprintf("register-tenant-handler %d vlan %d", vlanID, vlanID)
+	if output, err := r.suricataScFn(scCmd); err != nil {
+		return fmt.Errorf("failed to register handler for tenant %d with vlan %d: %v", vlanID, vlanID, string(output))
```

Comment:
If this fails, the tenant conf file still exists, the code won't register tenant handler again when retrying, leading to unrecoverable issue.

Reply:
Add code to clean config yaml and unregister tenant.